### PR TITLE
Add a staging notion to libobjects

### PR DESCRIPTION
**Pull request by maximedenes**
This change is the dual to #16056, where we introduced the same distinction on summaries.

Objects now take a stage at declaration time. We provide two instances of module operations and `Lib` (recording of session objects) maintenance and traversal, one for each stage.

This is a step towards #15409.

### CODE REVIEW

**Comment by maximedenes** [comment]
Like #16056, this PR is a piece of https://github.com/coq/coq/pull/15825, which is an experimental branch. The point is to avoid having to review massive changes in one go, while still getting a global picture by looking at the experimental branch.

File: checker/check.ml
Comment (maximedenes) [inline checker/check.ml:283]: One thing I was wondering: is there a procedure to follow these days, when changing the format of `.vo` files? (magic number,...)

File: library/global.ml
Comment (gares) [inline library/global.ml:62]: I tried to understand what this undocumented flag does, I even looked back at the PR which introduced it, but I don't get it. Here you use it, so maybe you understood it. Why is it needed and what does it do?

Comment (gares) [reply library/global.ml:62]: After some digging, here there is a need and a lurking bug unrelated to this PR:
- the need of restoring only a subset of the system state (summary), which seems legit
- a hack to reset/undo some system component even if it is not part of the installed state. This need/workaround dates 2000 by JCF, so I'm clueless what that is about. Clearly, if you restore a partial state you don't want to touch the other parts.

File: interp/modintern.ml
Comment (SkySkimmer) [inline interp/modintern.ml:109]: Why this change? the syntax looks like `with Definition fqid@{udecl} := c`

File: library/lib.ml
Comment (SkySkimmer) [inline library/lib.ml:98]: Please make non-catchall

File: library/lib.mli
Comment (SkySkimmer) [inline library/lib.mli:85]: Perhaps `type classified_objects` could be moved into this module type. Then the 2 separate instances of it would be nonconvertible and it would be clear in the type of `end_compilation` which part of the tuple is which stage.

File: library/lib.ml
Comment (SkySkimmer) [inline library/lib.ml:586]: If you move this before the functor and use it as the functor's type we get more errors in merlin while developping instead of having to run ocamlc

File: library/libobject.ml
Comment (SkySkimmer) [inline library/libobject.ml:148]: Shouldn't this be 
```suggestion
      object_stage = odecl.object_stage;
```
?
If named objects are not allowed to be synterp then add an assert instead of ignoring wrong declarations.

File: library/lib.mli
Comment (SkySkimmer) [inline library/lib.mli:79]: Why undeprecate? Is the Global one broken?

File: vernac/declaremods.ml
Comment (SkySkimmer) [inline vernac/declaremods.ml:1127]: Is this intended to be done before merging?

File: vernac/declaremods.ml
Comment (SkySkimmer) [inline vernac/declaremods.ml:848]: Is this TODO intended to be solved before merging?

File: vernac/declaremods.ml
Comment (SkySkimmer) [inline vernac/declaremods.ml:1144]: Is this TODO intended to be solved before merging?

File: library/lib.mli
Comment (maximedenes) [reply library/lib.mli:79]: The Global one requires a full environment, so it is not well suited to components which need to know whether they are in a section at parsing time (and such components exist).

File: interp/modintern.ml
Comment (maximedenes) [reply interp/modintern.ml]: Fixed, thanks!

File: library/lib.ml
Comment (maximedenes) [reply library/lib.ml]: Fixed, thanks!

File: vernac/declaremods.ml
Comment (maximedenes) [reply vernac/declaremods.ml]: Now done, thanks!

File: vernac/declaremods.ml
Comment (maximedenes) [reply vernac/declaremods.ml]: No, it was actually the same as @gares ' question. I removed the TODO.

File: vernac/declaremods.ml
Comment (maximedenes) [reply vernac/declaremods.ml]: No, it was actually the same as @gares ' question. I removed the TODO.

File: library/lib.ml
Comment (maximedenes) [reply library/lib.ml]: Done, thanks for the suggestion.

**Comment by Alizter** [comment]
Linter says:
```
No newline at end of file ./library/lib.mli!
```

File: library/lib.mli
Comment (Alizter) [inline library/lib.mli:175]:
```suggestion
  Nametab.object_prefix * Summary.frozen * classified_objects
  
```

File: library/lib.mli
Comment (maximedenes) [reply library/lib.mli:79]: Thanks, looking much better now:
```ocaml
val end_compilation : DirPath.t -> Nametab.object_prefix * Interp.classified_objects * Synterp.classified_objects
```

**Comment by maximedenes** [comment]
All green except for fiat-crypto, which I believe is a spurious failure.

@SkySkimmer does this PR now look good to you?

**Comment by SkySkimmer** [comment]
Interp and Synterp match (up to currying and `*` associativity)
~~~ocaml

module type InterpS = sig

  type declare_module_in
  type declare_module_out
  val declare_module : Id.t -> declare_module_in -> ModPath.t * declare_module_out

  type start_module_in
  type start_module_out
  val start_module : Lib.export -> Id.t -> start_module_in -> ModPath.t * start_module_out

  val end_module : unit -> ModPath.t

  type declare_modtype_in
  type declare_modtype_out
  val declare_modtype : Id.t -> declare_modtype_in -> ModPath.t * declare_modtype_out

  type start_modtype_in
  type start_modtype_out
  val start_modtype : Id.t -> start_modtype_in -> ModPath.t * start_modtype_out

  val end_modtype : unit -> ModPath.t

  val import_module : Libobject.open_filter -> export:Lib.export_flag -> ModPath.t -> unit

  val import_modules : export:Lib.export_flag -> (Libobject.open_filter * ModPath.t) list -> unit

  type register_library_in
  val register_library :
    library_name ->
    register_library_in ->
    unit

  type declare_include_in
  type declare_include_out
  val declare_include : declare_include_in -> declare_include_out

end
~~~
with the `*_out` of synterp being the `*_in` of interp.

I'm not sure if it's worth putting this in the code though.

**Comment by SkySkimmer** [comment]
Pushed small improvement

**Review by gares — APPROVED** [review]
I'm not an expert of this code, but I saw no problems either

**Comment by gares** [comment]
I can assign and click the merge button, but I'm not really in the teams responsible for the code paths which were touched.
Any volunteer?

**Review by ejgallego — DISMISSED** [review]
If I am not mistaken the design here means that the same meta-logical object may require two entries in the .vo file, I think this is the wrong API to expose, as it introduces a lot of duplication (as witnessed by this PR), it is not extensible to other traits by objects (which is a fundamental requirement), and contributes to possible technical debt as no static invariants about the separation are provided, and goes against recent proposals (c.f. https://github.com/coq/ceps/pull/65) to improve the libobject API for use from clients such as IDEs. It is also hard to judge the quality of the API until some real uses appear.

At the very least this PR needs to be discussed.

Right now, I'd propose to follow the path in the CEP: instead of duplicating the object for each possible trait it may implement, attach a (sealed) trait value for the syntactic action, that can be later specialized to provide static guarantees about the phase.

**Comment by maximedenes** [comment]
I'm afraid that on the engineering side, we clearly don't have the resources to pursue this work efficiently, if we need to repeatedly:
- discuss implicit "fundamental requirements"
- discuss the link between static invariants and technical debt
- block work when it is "hard to judge the quality"
- point to a CEP that should be followed even though it is more recent than this work (another CEP may pop up before the next rebase, I guess)

I tried many things to set up a way to make progress efficiently on UI implementations while you sort out related research questions, but failed in face of repeated request for changes in the direction of the engineering effort.

So I guess the most efficient way forward is simply that you take over this work. Judging from the past, it will save us *a lot* of communication cost. And don't get me wrong, I'm absolutely fine with this idea. Meanwhile, I'll switch to working on our CI, there's a lot to be done there too.

**Comment by gares** [comment]
@mattam82 was there a discussion about this direction or work in today's meeting?

**Comment by Alizter** [comment]
@gares It was not discussed AFAIK.

**Comment by maximedenes** [comment]
Reopening this because we (@ybertot and I) want to see this integrated. We are perfectly fine with having a next iteration to improve it (once traits land and if they subsume this, for example). But meanwhile, we want to make progress on the VsCoq front.

**Comment by gares** [comment]
@coqbot run full ci

**Comment by gares** [comment]
@coqbot bench

**Comment by gares** [comment]
@coqbot merge now

**Comment by Zimmi48** [comment]
It looks like this PR broke the build of Coq on `master` (probably a semantic conflict that was undetected by merging with a 9 days old CI).

**Comment by SkySkimmer** [comment]
build fix https://github.com/coq/coq/pull/16895

**Comment by gares** [comment]
Thanks @SkySkimmer

**Pull request merged**
